modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheEntryListenerHandler.java

```diff
@@ -0,0 +1,266 @@
+import java.util.function.Consumer;
+import javax.cache.Cache;
+import javax.cache.configuration.Factory;
+import javax.cache.event.CacheEntryEvent;
+import javax.cache.event.CacheEntryEventFilter;
+import javax.cache.event.CacheEntryUpdatedListener;
+import javax.cache.event.EventType;
+import org.apache.ignite.client.ClientChannelDisconnectListener;
+import org.apache.ignite.client.ClientException;
+import org.apache.ignite.internal.binary.GridBinaryMarshaller;
+import org.apache.ignite.internal.binary.streams.BinaryByteBufferInputStream;
+import org.apache.ignite.internal.binary.streams.BinaryInputStream;
+import org.apache.ignite.internal.binary.streams.BinaryOutputStream;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+/**
+ *
+ */
+public class ClientCacheEntryListenerHandler<K, V> implements NotificationListener, AutoCloseable {
+    /** "Keep binary" flag mask. */
+    private static final byte KEEP_BINARY_FLAG_MASK = 0x01;
+
+    /** */
+    private final Cache<K, V> jCacheAdapter;
+
+    /** */
+    private final ReliableChannel ch;
+
+    /** */
+    private final boolean keepBinary;
+
+    /** */
+    private final ClientUtils utils;
+
+    /** */
+    private final Consumer<ClientChannel> chCloseLsnr = this::onChannelClosed;
+
+    /** */
+    private volatile CacheEntryUpdatedListener<K, V> locLsnr;
+
+    /** */
+    private volatile ClientChannel clientCh;
+
+    /** */
+    private volatile long rsrcId;
+
+    /** */
+    ClientCacheEntryListenerHandler(
+        Cache<K, V> jCacheAdapter,
+        ReliableChannel ch,
+        ClientBinaryMarshaller marsh,
+        boolean keepBinary
+    ) {
+        this.jCacheAdapter = jCacheAdapter;
+        this.ch = ch;
+        this.keepBinary = keepBinary;
+        utils = new ClientUtils(marsh);
+    }
+
+    /**
+     * Send request to the server and start
+     */
+    public synchronized void startListen(
+        CacheEntryUpdatedListener<K, V> locLsnr,
+        Factory<? extends CacheEntryEventFilter<? super K, ? super V>> rmtFilterFactory,
+        int pageSize,
+        long timeInterval,
+        boolean includeExpired
+    ) {
+        if (clientCh != null)
+            throw new IllegalStateException("Listener was already started");
+
+        this.locLsnr = locLsnr;
+
+        Consumer<PayloadOutputChannel> qryWriter = payloadCh -> {
+            BinaryOutputStream out = payloadCh.out();
+
+            out.writeInt(ClientUtils.cacheId(jCacheAdapter.getName()));
+            out.writeByte(keepBinary ? KEEP_BINARY_FLAG_MASK : 0);
+            out.writeInt(pageSize);
+            out.writeLong(timeInterval);
+            out.writeBoolean(includeExpired);
+
+            if (rmtFilterFactory == null)
+                out.writeByte(GridBinaryMarshaller.NULL);
+            else {
+                utils.writeObject(out, rmtFilterFactory);
+                out.writeByte((byte)1); // Java platform
```

Comment:
`ClientPlatform.JAVA`

Reply:
`ClientPlatform' it's a server-side class, perhaps it's better to introduce some client-side constant and refactor `scanQuery` as well.

Reply:
> introduce some client-side constant

Sounds good to me. Let's avoid magic numbers.